Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -0,0 +1,188 @@
+
+namespace Ideal
+open Matrix
+
+variable {R : Type*} [Semiring R]
+         (n : Type*) [Fintype n] [DecidableEq n]
+
+/-- The left ideal of matrices with entries in `I ≤ R`. -/
+def matricesOver (I : Ideal R) : Ideal (Matrix n n R) where
+  carrier := { M | ∀ i j, M i j ∈ I }
+  add_mem' ha hb i j := I.add_mem (ha i j) (hb i j)
+  zero_mem' _ _ := I.zero_mem
+  smul_mem' M N hN := by
+    intro i j
+    rw [smul_eq_mul, mul_apply]
+    apply sum_mem
+    intro k _
+    apply I.mul_mem_left _ (hN k j)
+
+@[simp]
+theorem mem_matricesOver (I : Ideal R) (M : Matrix n n R) :
+    M ∈ I.matricesOver n ↔ ∀ i j, M i j ∈ I := by rfl
+
+def matricesOver_mono {I J : Ideal (Matrix n n R)} :
+    I ≤ J → I.matricesOver n ≤ J.matricesOver n :=
+  fun IJ _ MI i j => IJ (MI i j)
+
+@[simp]
+theorem matricesOver_bot : (⊥ : Ideal R).matricesOver n = ⊥ := by
+  ext M
+  simp only [mem_matricesOver, mem_bot]
+  constructor
+  · intro H; ext; apply H
+  · intro H; simp [H]
+
+@[simp]
+theorem matricesOver_top : (⊤ : Ideal R).matricesOver n = ⊤ := by
+  ext; simp
+
+end Ideal
+
+/-! ### Two-sided ideals in a matrix ring -/
+
+namespace TwoSidedIdeal
+open Matrix
+
+variable {R : Type*} [Ring R]
+         (n : Type*) [Fintype n]
+
+/-- The two-sided ideal of matrices with entries in `I ≤ R`. -/
+def matricesOver (I : TwoSidedIdeal R) : TwoSidedIdeal (Matrix n n R) :=
+  TwoSidedIdeal.mk' { M | ∀ i j, M i j ∈ I }
+    (fun _ _ => I.zero_mem)
+    (fun ha hb i j => I.add_mem (ha i j) (hb i j))
+    (fun ha i j => I.neg_mem (ha i j))
+    (fun ha i j => by
+      rw [mul_apply]
+      apply sum_mem
+      intro k _
+      apply I.mul_mem_left _ _ (ha k j))
+    (fun ha i j => by
+      rw [mul_apply]
+      apply sum_mem
+      intro k _
+      apply I.mul_mem_right _ _ (ha i k))
+
+@[simp]
+lemma mem_matricesOver (I : TwoSidedIdeal R) (M : Matrix n n R) :
+    M ∈ I.matricesOver n ↔ ∀ i j, M i j ∈ I := by
+  simp [matricesOver]
+
+def matricesOver_mono {I J : TwoSidedIdeal R} :
+    I ≤ J → I.matricesOver n ≤ J.matricesOver n :=
+  fun IJ _ MI i j => IJ (MI i j)
+
+@[simp]
+theorem matricesOver_bot : (⊥ : TwoSidedIdeal R).matricesOver n = ⊥ := by
+  ext M
+  simp only [mem_matricesOver, mem_bot]
+  constructor
+  · intro H; ext; apply H
+  · intro H; simp [H]
+
+@[simp]
+theorem matricesOver_top : (⊤ : TwoSidedIdeal R).matricesOver n = ⊤ := by
+  ext; simp
+
+theorem asIdeal_matricesOver [DecidableEq n] (I : TwoSidedIdeal R) :
+    asIdeal (I.matricesOver n) = (asIdeal I).matricesOver n := by
+  ext; simp
+
+variable {n : Type*} [Fintype n] [DecidableEq n]
+
+/--
+Two-sided ideals in `R` correspond bijectively to those in `Mₙ(R)`.
+Given an ideal `I ≤ R`, we send it to `Mₙ(I)`.
+Given an ideal `J ≤ Mₙ(R)`, we send it to `{x₀₀ | x ∈ J}`.
+-/
+@[simps]
+def equivMatricesOver (o : n) : TwoSidedIdeal R ≃ TwoSidedIdeal (Matrix n n R) where
+  toFun I := I.matricesOver n
+  invFun J := TwoSidedIdeal.mk'
+    { x : R | ∃ N ∈ J, x = N o o }
```

Comment:
Does this proof rely on this being `N o o`, or would `N i j` work too?

Reply:
Any `i j` can work. Certainly the underlying construction of moving a particular entry to a specified position by multiplying with basis matrices is useful, though I'm not sure how useful it is when wrapped in this equivalence. Nevertheless I generalized it.